extensions/amp-story/1.0/test/test-amp-story.js

```diff
@@ -540,6 +540,54 @@ describes.realWin(
         });
     });
 
+    it('should deduplicate amp-story-page ids', async () => {
+      createPages(story.element, 6, [
+        'cover',
+        'page-1',
+        'cover',
+        'page-1',
+        'page-1',
+        'page-2',
+      ]);
+
+      allowConsoleError(() => story.buildCallback());
+      await story.layoutCallback();
+
+      const pages = story.element.querySelectorAll('amp-story-page');
+      const pageIds = Array.prototype.map.call(pages, page => page.id);
+      expect(pageIds).to.deep.equal([
+        'cover',
+        'page-1',
+        'cover__1',
+        'page-1__1',
+        'page-1__2',
+        'page-2',
+      ]);
+    });
+
+    it('should deduplicate amp-story-page ids and cache them', async () => {
+      createPages(story.element, 6, [
+        'cover',
+        'page-1',
+        'cover',
+        'page-1',
+        'page-1',
+        'page-2',
+      ]);
+
+      allowConsoleError(() => story.buildCallback());
+      await story.layoutCallback();
+
+      expect(story.storeService_.get(StateProperty.PAGE_IDS)).to.deep.equal([
+        'cover',
```

Comment:
It's due to the condition set in the `while ((index = pageIds.indexOf(pageIds[i], i + 1)) !== -1) {` of line 531 in amp-story.js.

the `index` will take the value of the first duplicate it finds after the current `i` element. Thus, the  current element remains unchanged.

Perhaps there is a way to make that loop a bit easier to read.

Reply:
The second parameter of `indexOf` gives a start index. Goal was to reduce the O(n2) complexity since indexOf will loop over the array.
I was trying to do as little updates as possible, which is why it's keeping the first id as `cover` and not `cover__0`. I thought it made more product sense (ie: keep what the publisher wrote as much as possible), but it also makes the code easier and faster.

Reply:
Actually I rewrote the loop to make it easier to read and O(n), thanks Enrique.